### PR TITLE
use lseek(),read() and write() to emulate pread() and pwrite()

### PR DESCRIPTION
**Pull request by BaoshanPang**
r? @alexcrichton

### CODE REVIEW

**Comment by BaoshanPang** [comment]
cc @n-salim

**Comment by BaoshanPang** [comment]
cc @UmeshKalappa

File: src/libstd/sys/vxworks/fd.rs [hunk 1]
```diff
@@ -58,19 +58,13 @@ impl FileDesc {
...
-        }
-
         unsafe {
-            cvt_pread(self.fd,
-                buf.as_mut_ptr() as *mut c_void,
-                cmp::min(buf.len(), max_len()),
-                offset as i64)
-            .map(|n| n as usize)
+            let orig_off = libc::lseek(self.fd, 0, libc::SEEK_CUR);
+            libc::lseek(self.fd, offset as i64, libc::SEEK_SET);
```
Comment (nagisa) [inline src/libstd/sys/vxworks/fd.rs]: Must check errors returned by `lseek`.

File: src/libstd/sys/vxworks/fd.rs [hunk 2]
```diff
@@ -58,19 +58,13 @@ impl FileDesc {
...
+            let ret = libc::read(self.fd, buf.as_mut_ptr() as *mut c_void,
```
Comment (nagisa) [inline src/libstd/sys/vxworks/fd.rs]: Similarly, must check the errors returned by `read` call.

File: src/libstd/sys/vxworks/fd.rs [hunk 3]
```diff
@@ -93,19 +87,13 @@ impl FileDesc {
...
-
         unsafe {
-            cvt_pwrite(self.fd,
-                buf.as_ptr() as *const c_void,
-                cmp::min(buf.len(), max_len()),
-                offset as i64)
-                .map(|n| n as usize)
+            let orig_off = libc::lseek(self.fd, 0, libc::SEEK_CUR);
+            libc::lseek(self.fd, offset as i64, libc::SEEK_SET);
+            let ret = libc::write(self.fd, buf.as_ptr() as *const c_void,
```
Comment (nagisa) [inline src/libstd/sys/vxworks/fd.rs]: Similarly, must check the errors returned by the `write` call.

File: src/libstd/sys/vxworks/fd.rs [hunk 4]
```diff
@@ -58,19 +58,13 @@ impl FileDesc {
     }
 
     pub fn read_at(&self, buf: &mut [u8], offset: u64) -> io::Result<usize> {
```
Comment (nagisa) [inline src/libstd/sys/vxworks/fd.rs]: This function is expected to be thread-safe, this implementation of it is not. To implement this function you will need locking somewhere.

File: src/libstd/sys/vxworks/fd.rs [hunk 5]
```diff
@@ -93,19 +87,13 @@ impl FileDesc {
     }
 
     pub fn write_at(&self, buf: &[u8], offset: u64) -> io::Result<usize> {
```
Comment (nagisa) [inline src/libstd/sys/vxworks/fd.rs]: This function is expected to be thread-safe, this implementation of it is not. To implement this function you will need locking somewhere.

**Review by nagisa — CHANGES_REQUESTED** [review]
More motivation and explanation would be great. 

A cursory search around the internet suggests that the `pread`/`pwrite` are indeed not implemented by WxWorks, but this is not something I, or @alexcrichton, should be doing.

**Comment by n-salim** [comment]
@BaoshanPang Sorry to chime in on this so late: I don't think we can usefully implement pread and pwrite without VxWorks kernel support. 

pread/pwrite are used to access different parts of a file concurrently. From the man page:

"The pwrite() function shall be equivalent to write(), except that it writes into a given position without 
changing the file pointer. "

I take that to mean that not only can I have multiple pread/pwrites to different parts of a file concurrently but that they should also not interfere with an ordinary "write". 

I.e. if the implementation mucks around with the file offset then it must be done in a way that is not "observable" by any other thread.

Yes, we've seen non-locking "implementations" based on lseek used to mock-up pread/pwrite for specific applications, but that isn't a robust approach for general use. 

Apparently older versions of Linux also "emulate" pread/pwrite in libc. However I believe those implementations are also not thread safe.

While it might be possible in principle to implement pread/pwrite in a thread safe way entirely within the user space library, I can't see how to do that in a way that interacts properly with concurrent uses of write and/or lseek without adding an additional locking penalty on all of the low-level APIs.

So I'd prefer to leave these APIs as unimplemented for now (and in the meantime we can work with the OS team to try to add proper kernel support for pread/pwrite).

**Comment by nagisa** [comment]
One way to make this work out could be to `dup` the file descriptor, and then do whatever `lseek`/`read`/`write` dance that is necessary, but it does not look like WxWorks supports that either.

**Comment by n-salim** [comment]
> One way to make this work out could be to `dup` the file descriptor, and then do whatever `lseek`/`read`/`write` dance that is necessary, but it does not look like WxWorks supports that either.

Assuming you're referring to the underlying POSIX libc API "dup" and not some mechanism in the Rust standard library: VxWorks does actually support it, but I'm not sure that solves the problem, because as I understand it, the "new" file descriptor is supposed to refer to the same underlying descriptor - in particular they both share a file offset, so lseek on one of the descriptors will affect both of them.

**Comment by nagisa** [comment]
Ah, you are right.

**Comment by BaoshanPang** [comment]
@nagisa 

Thanks for helping on reviewing this. but as Salim suggested this should be implemented in vxWorks kernel, so I will close this PR.

**Comment by BaoshanPang** [comment]
@n-salim 
I will close this PR, and let OS team fix it in kernel.

**Pull request closed without merging**